File: back-end/src/services/admin.service.js
const db = require('../config/db');
const userModel = require('../models/user.model');
const auditLogModel = require('../models/auditLog.model');
const consentLogModel = require('../models/consentLog.model');
const recommendationModel = require('../models/recommendation.model');
const venueModel = require('../models/venue.model');
const photoCacheModel = require('../models/photoCache.model');

const DELETION_GRACE_DAYS = Number(process.env.ADMIN_DELETION_GRACE_DAYS) || 14;

// Routes every admin action through here so it lands in admin_audit_log; logging failures never block the action.
const logAdminAction = async (req, action, target, details) => {
  try {
    await auditLogModel.insert({
      adminId: req.user.userId,
      action,
      targetUserId: target ? target.user_id : null,
      targetEmail: target ? target.email : null,
      details,
    });
  } catch (err) {
    console.error('Failed to write admin audit log entry', err);
  }
};

// FR1.7 — listing users is reserved for deletion requests and security threats, not general browsing.
exports.listUsers = async (req, res) => {
  try {
    const result = await userModel.findAllForAdmin();
    return res.status(200).json({ users: result.rows });
  } catch (err) {
    console.error('Admin list users error', err);
    return res.status(500).json({ message: 'Failed to load users', error: err.message });
  }
};

// Only view that returns the unmasked mobile_number, so every call is audit-logged.
exports.getUser = async (req, res) => {
  try {
    const result = await userModel.findByIdForAdmin(req.params.userId);
    if (result.rows.length === 0) {
      return res.status(404).json({ message: 'User not found' });
    }
    await logAdminAction(req, 'view_profile', result.rows[0]);
    return res.status(200).json({ user: result.rows[0] });
  } catch (err) {
    console.error('Admin get user error', err);
    return res.status(500).json({ message: 'Failed to load user', error: err.message });
  }
};

// FR1.7 — deactivate. Reversible: a deactivated account cannot log in
// (see auth.service.js) but its data is preserved.
exports.deactivateUser = async (req, res) => {
  const { userId } = req.params;

  if (userId === req.user.userId) {
    return res.status(400).json({ message: 'An admin cannot deactivate their own account' });
  }

  try {
    const existing = await userModel.findByIdForAdmin(userId);
    if (existing.rows.length === 0) {
      return res.status(404).json({ message: 'User not found' });
    }

    const result = await userModel.setActiveStatus(userId, false);
    await logAdminAction(req, 'deactivate', result.rows[0]);
    return res.status(200).json({ message: 'User account deactivated', user: result.rows[0] });
  } catch (err) {
    console.error('Admin deactivate user error', err);
    return res.status(500).json({ message: 'Failed to deactivate user', error: err.message });
  }
};

// Symmetric counterpart to deactivation — restores login access without
// requiring a destructive delete-and-recreate cycle.
exports.reactivateUser = async (req, res) => {
  const { userId } = req.params;

  try {
    const existing = await userModel.findByIdForAdmin(userId);
    if (existing.rows.length === 0) {
      return res.status(404).json({ message: 'User not found' });
    }

    const result = await userModel.setActiveStatus(userId, true);
    await logAdminAction(req, 'reactivate', result.rows[0]);
    return res.status(200).json({ message: 'User account reactivated', user: result.rows[0] });
  } catch (err) {
    console.error('Admin reactivate user error', err);
    return res.status(500).json({ message: 'Failed to reactivate user', error: err.message });
  }
};

// PDPA erasure with a recovery window: soft-deletes now, purgeUser removes the row for good after the grace period.
exports.deleteUser = async (req, res) => {
  const { userId } = req.params;

  if (userId === req.user.userId) {
    return res.status(400).json({ message: 'An admin cannot delete their own account' });
  }

  try {
    const existing = await userModel.findByIdForAdmin(userId);
    if (existing.rows.length === 0) {
      return res.status(404).json({ message: 'User not found' });
    }

    const result = await userModel.requestDeletion(userId);
    await logAdminAction(req, 'request_deletion', result.rows[0], { graceDays: DELETION_GRACE_DAYS });
    return res.status(200).json({ message: 'Deletion requested; account deactivated pending purge', user: result.rows[0] });
  } catch (err) {
    console.error('Admin request deletion error', err);
    return res.status(500).json({ message: 'Failed to request deletion', error: err.message });
  }
};

// Reverses a pending deletion request within the grace window.
exports.cancelDeletion = async (req, res) => {
  const { userId } = req.params;

  try {
    const existing = await userModel.findByIdForAdmin(userId);
    if (existing.rows.length === 0) {
      return res.status(404).json({ message: 'User not found' });
    }
    if (!existing.rows[0].deletion_requested_at) {
      return res.status(400).json({ message: 'This user has no pending deletion request' });
    }

    const result = await userModel.cancelDeletionRequest(userId);
    await logAdminAction(req, 'cancel_deletion', result.rows[0]);
    return res.status(200).json({ message: 'Deletion request cancelled', user: result.rows[0] });
  } catch (err) {
    console.error('Admin cancel deletion error', err);
    return res.status(500).json({ message: 'Failed to cancel deletion request', error: err.message });
  }
};

// FK cascades (see 024_admin_user_management.sql) wipe all related rows in one statement; blocked until the grace period elapses unless ?force=true.
exports.purgeUser = async (req, res) => {
  const { userId } = req.params;
  const force = req.query.force === 'true';

  if (userId === req.user.userId) {
    return res.status(400).json({ message: 'An admin cannot delete their own account' });
  }

  try {
    const existing = await userModel.findByIdForAdmin(userId);
    if (existing.rows.length === 0) {
      return res.status(404).json({ message: 'User not found' });
    }

    const target = existing.rows[0];
    if (!target.deletion_requested_at) {
      return res.status(400).json({ message: 'This user has no pending deletion request. Request deletion first.' });
    }

    const requestedAt = new Date(target.deletion_requested_at);
    const graceElapsedMs = Date.now() - requestedAt.getTime();
    const graceElapsed = graceElapsedMs >= DELETION_GRACE_DAYS * 24 * 60 * 60 * 1000;

    if (!graceElapsed && !force) {
      const daysLeft = Math.ceil((DELETION_GRACE_DAYS * 24 * 60 * 60 * 1000 - graceElapsedMs) / (24 * 60 * 60 * 1000));
      return res.status(400).json({
        message: `Grace period has not elapsed (${daysLeft} day(s) remaining). Pass ?force=true to purge immediately.`,
      });
    }

    // Logged before the row is gone — afterwards target_user_id FK nulls out,
    // but target_email survives as the record of what was purged and why.
    await logAdminAction(req, 'purge_user', target, { forced: !graceElapsed && force });

    const result = await userModel.deleteUser(userId);
    return res.status(200).json({ message: 'User account permanently deleted', user: result.rows[0] });
  } catch (err) {
    console.error('Admin purge user error', err);
    return res.status(500).json({ message: 'Failed to purge user', error: err.message });
  }
};

// PDPA data subject access request: dumps everything the app holds on one user as JSON.
exports.exportUserData = async (req, res) => {
  const { userId } = req.params;

  try {
    const userResult = await userModel.findByIdForAdmin(userId);
    if (userResult.rows.length === 0) {
      return res.status(404).json({ message: 'User not found' });
    }
    const profile = userResult.rows[0];

    const [expenses, budgets, savingGoals, wishlist, ocrReceipts, recommendationLogs, incomeLog, consentHistory] =
      await Promise.all([
        db.query('SELECT * FROM expenses WHERE user_id = $1', [userId]),
        db.query('SELECT * FROM budgets WHERE user_id = $1', [userId]),
        db.query('SELECT * FROM saving_goals WHERE user_id = $1', [userId]),
        db.query('SELECT * FROM wishlist WHERE user_id = $1', [userId]),
        db.query('SELECT * FROM ocr_receipts WHERE user_id = $1', [userId]),
        db.query('SELECT * FROM recommendation_log WHERE user_id = $1', [userId]),
        db.query('SELECT * FROM income_log WHERE user_id = $1', [userId]),
        consentLogModel.findForUser(userId),
      ]);

    await logAdminAction(req, 'export_data', profile, {
      recordCounts: {
        expenses: expenses.rows.length,
        budgets: budgets.rows.length,
        savingGoals: savingGoals.rows.length,
        wishlist: wishlist.rows.length,
        ocrReceipts: ocrReceipts.rows.length,
        recommendationLogs: recommendationLogs.rows.length,
        incomeLog: incomeLog.rows.length,
        consentHistory: consentHistory.rows.length,
      },
    });

    const exportPayload = {
      profile,
      expenses: expenses.rows,
      budgets: budgets.rows,
      savingGoals: savingGoals.rows,
      wishlist: wishlist.rows,
      ocrReceipts: ocrReceipts.rows,
      recommendationLogs: recommendationLogs.rows,
      incomeLog: incomeLog.rows,
      consentHistory: consentHistory.rows,
      exportedAt: new Date().toISOString(),
    };

    res.setHeader('Content-Disposition', `attachment; filename="user-${userId}-export.json"`);
    return res.status(200).json(exportPayload);
  } catch (err) {
    console.error('Admin export user data error', err);
    return res.status(500).json({ message: 'Failed to export user data', error: err.message });
  }
};

// PDPA Notice & Choice — evidence of when this user granted/withdrew consent.
exports.getConsentHistory = async (req, res) => {
  try {
    const result = await consentLogModel.findForUser(req.params.userId);
    return res.status(200).json({ consentHistory: result.rows });
  } catch (err) {
    console.error('Admin get consent history error', err);
    return res.status(500).json({ message: 'Failed to load consent history', error: err.message });
  }
};

// PDPA Accountability — the audit trail itself, for the admin panel.
exports.getAuditLog = async (req, res) => {
  try {
    const result = await auditLogModel.findRecent(200);
    return res.status(200).json({ auditLog: result.rows });
  } catch (err) {
    console.error('Admin get audit log error', err);
    return res.status(500).json({ message: 'Failed to load audit log', error: err.message });
  }
};

// context_snapshot holds raw GPS coords with no reason to keep indefinitely; admin-triggered since there's no job scheduler.
exports.purgeRecommendationLogs = async (req, res) => {
  const olderThanDays = Number(req.query.olderThanDays);
  if (!Number.isFinite(olderThanDays) || olderThanDays < 0) {
    return res.status(400).json({ message: 'olderThanDays must be a non-negative number' });
  }

  try {
    const result = await recommendationModel.purgeOlderThan(olderThanDays);
    await logAdminAction(req, 'purge_recommendation_logs', null, {
      olderThanDays,
      deletedCount: result.rows.length,
    });
    return res.status(200).json({ message: 'Old recommendation logs purged', deletedCount: result.rows.length });
  } catch (err) {
    console.error('Admin purge recommendation logs error', err);
    return res.status(500).json({ message: 'Failed to purge recommendation logs', error: err.message });
  }
};

// Guest login mints a new row every time with no reuse, so abandoned guests pile up; unlike log purges this deletes identities, so the audit entry records the purged emails.
exports.purgeStaleGuests = async (req, res) => {
  const olderThanDays = Number(req.query.olderThanDays);
  if (!Number.isFinite(olderThanDays) || olderThanDays < 0) {
    return res.status(400).json({ message: 'olderThanDays must be a non-negative number' });
  }

  try {
    const result = await userModel.purgeStaleGuests(olderThanDays);
    const purgedEmails = result.rows.map((row) => row.email);
    await logAdminAction(req, 'purge_stale_guests', null, {
      olderThanDays,
      deletedCount: result.rows.length,
      purgedEmails,
    });
    return res.status(200).json({ message: 'Stale guest accounts purged', deletedCount: result.rows.length });
  } catch (err) {
    console.error('Admin purge stale guests error', err);
    return res.status(500).json({ message: 'Failed to purge stale guest accounts', error: err.message });
  }
};

// venue_cache has no automatic expiry, so stale entries just accumulate; admin-triggered since there's no job scheduler.
exports.purgeVenueCache = async (req, res) => {
  const olderThanDays = Number(req.query.olderThanDays);
  if (!Number.isFinite(olderThanDays) || olderThanDays < 0) {
    return res.status(400).json({ message: 'olderThanDays must be a non-negative number' });
  }

  try {
    const result = await venueModel.purgeStale(olderThanDays);
    await logAdminAction(req, 'purge_venue_cache', null, {
      olderThanDays,
      deletedCount: result.rows.length,
    });
    return res.status(200).json({ message: 'Stale venue cache entries purged', deletedCount: result.rows.length });
  } catch (err) {
    console.error('Admin purge venue cache error', err);
    return res.status(500).json({ message: 'Failed to purge venue cache', error: err.message });
  }
};

// photo_cache stores raw image bytes (see photoCache.model.js#purgeStale),
// so unbounded growth here is a storage-size concern, not just row count.
exports.purgePhotoCache = async (req, res) => {
  const olderThanDays = Number(req.query.olderThanDays);
  if (!Number.isFinite(olderThanDays) || olderThanDays < 0) {
    return res.status(400).json({ message: 'olderThanDays must be a non-negative number' });
  }

  try {
    const result = await photoCacheModel.purgeStale(olderThanDays);
    await logAdminAction(req, 'purge_photo_cache', null, {
      olderThanDays,
      deletedCount: result.rows.length,
    });
    return res.status(200).json({ message: 'Stale photo cache entries purged', deletedCount: result.rows.length });
  } catch (err) {
    console.error('Admin purge photo cache error', err);
    return res.status(500).json({ message: 'Failed to purge photo cache', error: err.message });
  }
};

File: front-end/lib/services/admin_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

import '../core/constants/api_endpoints.dart';
import 'auth_service.dart';

/// A row from GET /api/admin/users. password_hash is never returned by the
/// backend (see back-end/src/models/user.model.js findAllForAdmin).
///
/// mobile_number is masked in the list response (PDPA data-minimization —
/// see findAllForAdmin) — only GET /api/admin/users/:id returns it in full,
/// which is why this same class is reused for both endpoints rather than
/// having a separate "detail" type: the shape is identical, only how much
/// of mobile_number is visible differs server-side.
class AdminUser {
  final String userId;
  final String email;
  final String username;
  final String role;
  final bool isGuest;
  final bool isActive;
  final bool isLocked;
  final String? mobileNumber;
  final DateTime? deactivatedAt;
  final DateTime? deletionRequestedAt;
  final DateTime createdAt;

  const AdminUser({
    required this.userId,
    required this.email,
    required this.username,
    required this.role,
    required this.isGuest,
    required this.isActive,
    required this.isLocked,
    required this.mobileNumber,
    required this.deactivatedAt,
    required this.deletionRequestedAt,
    required this.createdAt,
  });

  factory AdminUser.fromJson(Map<String, dynamic> json) {
    return AdminUser(
      userId: json['user_id'] as String,
      email: json['email'] as String,
      username: json['username'] as String,
      role: json['role'] as String,
      isGuest: json['is_guest'] as bool,
      isActive: json['is_active'] as bool,
      isLocked: json['is_locked'] as bool,
      mobileNumber: json['mobile_number'] as String?,
      deactivatedAt: json['deactivated_at'] != null ? DateTime.parse(json['deactivated_at'] as String) : null,
      deletionRequestedAt:
          json['deletion_requested_at'] != null ? DateTime.parse(json['deletion_requested_at'] as String) : null,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }
}

/// A row from GET /api/admin/users/:id/consent-history (consent_log table).
class ConsentLogEntry {
  final String consentType;
  final bool granted;
  final DateTime changedAt;

  const ConsentLogEntry({required this.consentType, required this.granted, required this.changedAt});

  factory ConsentLogEntry.fromJson(Map<String, dynamic> json) {
    return ConsentLogEntry(
      consentType: json['consent_type'] as String,
      granted: json['granted'] as bool,
      changedAt: DateTime.parse(json['changed_at'] as String),
    );
  }
}

/// A row from GET /api/admin/audit-log (admin_audit_log table) — the
/// accountability trail of admin actions against user records.
class AdminAuditEntry {
  final String action;
  final String? targetEmail;
  final Map<String, dynamic> details;
  final DateTime createdAt;

  const AdminAuditEntry({
    required this.action,
    required this.targetEmail,
    required this.details,
    required this.createdAt,
  });

  factory AdminAuditEntry.fromJson(Map<String, dynamic> json) {
    return AdminAuditEntry(
      action: json['action'] as String,
      targetEmail: json['target_email'] as String?,
      details: (json['details'] as Map<String, dynamic>?) ?? const {},
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }
}

/// Service layer for FR1.7 (admin account management: view/deactivate/delete).
/// Every call requires a JWT for an account with role='admin' — the backend
/// enforces this via admin.middleware.js regardless of what the client sends.
class AdminService {
  final _authService = AuthService();

  Future<Map<String, String>> _authHeaders() async {
    final token = await _authService.getToken();
    return {
      'Content-Type': 'application/json',
      if (token != null) 'Authorization': 'Bearer $token',
    };
  }

  /// GET /api/admin/users
  Future<List<AdminUser>> fetchUsers() async {
    try {
      final response = await http.get(Uri.parse(ApiEndpoints.adminUsers()), headers: await _authHeaders());
      final body = jsonDecode(response.body) as Map<String, dynamic>;

      if (response.statusCode == 200) {
        return (body['users'] as List<dynamic>)
            .map((e) => AdminUser.fromJson(e as Map<String, dynamic>))
            .toList();
      } else {
        throw Exception(body['message'] ?? 'Failed to load users');
      }
    } catch (e) {
      throw Exception('Error loading users: $e');
    }
  }

  /// GET /api/admin/users/:id — full (unmasked) profile. Audit-logged
  /// server-side as a "view_profile" action every time it's called.
  Future<AdminUser> fetchUser(String userId) async {
    final response = await http.get(Uri.parse(ApiEndpoints.adminDeleteUser(userId)), headers: await _authHeaders());
    final body = jsonDecode(response.body) as Map<String, dynamic>;
    if (response.statusCode != 200) {
      throw Exception(body['message'] ?? 'Failed to load user (status ${response.statusCode})');
    }
    return AdminUser.fromJson(body['user'] as Map<String, dynamic>);
  }

  /// PATCH /api/admin/users/:id/deactivate
  Future<void> deactivateUser(String userId) => _patch(ApiEndpoints.adminDeactivateUser(userId));

  /// PATCH /api/admin/users/:id/reactivate
  Future<void> reactivateUser(String userId) => _patch(ApiEndpoints.adminReactivateUser(userId));

  /// PATCH /api/admin/users/:id/cancel-deletion — reverses a pending
  /// deletion request within the grace period.
  Future<void> cancelDeletion(String userId) => _patch(ApiEndpoints.adminCancelDeletion(userId));

  Future<void> _patch(String url) async {
    final response = await http.patch(Uri.parse(url), headers: await _authHeaders());
    if (response.statusCode != 200) {
      final body = jsonDecode(response.body) as Map<String, dynamic>;
      throw Exception(body['message'] ?? 'Request failed (status ${response.statusCode})');
    }
  }

  /// DELETE /api/admin/users/:id — soft: requests deletion (deactivates,
  /// starts the grace-period clock). Permanent removal only happens via
  /// [purgeUser] once that period elapses (PDPA erasure with a
  /// recoverability window, FR1.7).
  Future<void> deleteUser(String userId) async {
    final response = await http.delete(Uri.parse(ApiEndpoints.adminDeleteUser(userId)), headers: await _authHeaders());
    if (response.statusCode != 200) {
      final body = jsonDecode(response.body) as Map<String, dynamic>;
      throw Exception(body['message'] ?? 'Failed to request deletion (status ${response.statusCode})');
    }
  }

  /// DELETE /api/admin/users/:id/purge — permanent. Rejected with a 400
  /// (surfaced as an Exception here) until the grace period has elapsed,
  /// unless [force] is set.
  Future<void> purgeUser(String userId, {bool force = false}) async {
    final response = await http.delete(
      Uri.parse(ApiEndpoints.adminPurgeUser(userId, force: force)),
      headers: await _authHeaders(),
    );
    if (response.statusCode != 200) {
      final body = jsonDecode(response.body) as Map<String, dynamic>;
      throw Exception(body['message'] ?? 'Failed to purge user (status ${response.statusCode})');
    }
  }

  /// GET /api/admin/users/:id/export — a DSAR export of everything the app
  /// holds on this user. Returned as pretty-printed JSON text for display;
  /// the caller doesn't need it as a parsed object.
  Future<String> exportUser(String userId) async {
    final response = await http.get(Uri.parse(ApiEndpoints.adminExportUser(userId)), headers: await _authHeaders());
    if (response.statusCode != 200) {
      final body = jsonDecode(response.body) as Map<String, dynamic>;
      throw Exception(body['message'] ?? 'Failed to export user data (status ${response.statusCode})');
    }
    final decoded = jsonDecode(response.body);
    return const JsonEncoder.withIndent('  ').convert(decoded);
  }

  /// GET /api/admin/users/:id/consent-history
  Future<List<ConsentLogEntry>> fetchConsentHistory(String userId) async {
    final response =
        await http.get(Uri.parse(ApiEndpoints.adminConsentHistory(userId)), headers: await _authHeaders());
    final body = jsonDecode(response.body) as Map<String, dynamic>;
    if (response.statusCode != 200) {
      throw Exception(body['message'] ?? 'Failed to load consent history (status ${response.statusCode})');
    }
    return (body['consentHistory'] as List<dynamic>)
        .map((e) => ConsentLogEntry.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  /// GET /api/admin/audit-log — the accountability trail of admin actions.
  Future<List<AdminAuditEntry>> fetchAuditLog() async {
    final response = await http.get(Uri.parse(ApiEndpoints.adminAuditLog()), headers: await _authHeaders());
    final body = jsonDecode(response.body) as Map<String, dynamic>;
    if (response.statusCode != 200) {
      throw Exception(body['message'] ?? 'Failed to load audit log (status ${response.statusCode})');
    }
    return (body['auditLog'] as List<dynamic>).map((e) => AdminAuditEntry.fromJson(e as Map<String, dynamic>)).toList();
  }

  /// DELETE /api/admin/recommendation-logs/purge — retention purge of raw
  /// GPS coordinates stored by the food-recommendation feature. Returns the
  /// number of rows deleted.
  Future<int> purgeRecommendationLogs(int olderThanDays) async {
    final response = await http.delete(
      Uri.parse(ApiEndpoints.adminPurgeRecommendationLogs(olderThanDays)),
      headers: await _authHeaders(),
    );
    final body = jsonDecode(response.body) as Map<String, dynamic>;
    if (response.statusCode != 200) {
      throw Exception(body['message'] ?? 'Failed to purge recommendation logs (status ${response.statusCode})');
    }
    return body['deletedCount'] as int;
  }

  /// DELETE /api/admin/guests/purge — PDPA storage-limitation cleanup for
  /// "Continue as Guest" accounts abandoned past the given inactivity
  /// threshold (see user.model.js purgeStaleGuests for how "inactive" is
  /// computed). Returns the number of accounts deleted.
  Future<int> purgeStaleGuests(int olderThanDays) async {
    final response = await http.delete(
      Uri.parse(ApiEndpoints.adminPurgeStaleGuests(olderThanDays)),
      headers: await _authHeaders(),
    );
    final body = jsonDecode(response.body) as Map<String, dynamic>;
    if (response.statusCode != 200) {
      throw Exception(body['message'] ?? 'Failed to purge stale guest accounts (status ${response.statusCode})');
    }
    return body['deletedCount'] as int;
  }
}
